brigade/plugins/tasks/networking/netmiko_tasks.py

```diff
@@ -0,0 +1,161 @@
+}
+
+
+def get_template_dir():
+    """Find and return the ntc-templates/templates dir."""
+    try:
+        template_dir = os.environ['NET_TEXTFSM']
+        index = os.path.join(template_dir, 'index')
+        if not os.path.isfile(index):
+            # Assume only base ./ntc-templates specified
+            template_dir = os.path.join(template_dir, 'templates')
+    except KeyError:
+        # Construct path ~/ntc-templates/templates
+        home_dir = os.path.expanduser("~")
+        template_dir = os.path.join(home_dir, 'ntc-templates', 'templates')
+
+    index = os.path.join(template_dir, 'index')
+    if not os.path.isdir(template_dir) or not os.path.isfile(index):
+        msg = """
+Valid ntc-templates not found, please install https://github.com/networktocode/ntc-templates
+and then set the NET_TEXTFSM environment variable to point to the ./ntc-templates/templates
+directory."""
+        raise ValueError(msg)
+    return template_dir
+
+
+def get_structured_data(raw_output, platform, command):
+    """Convert raw CLI output to structured data using TextFSM template."""
+    template_dir = get_template_dir()
+    index_file = os.path.join(template_dir, 'index')
+    textfsm_obj = clitable.CliTable(index_file, template_dir)
+    attrs = {'Command': command, 'Platform': platform}
+    try:
+        # Parse output through template
+        textfsm_obj.ParseCmd(raw_output, attrs)
+        return clitable_to_dict(textfsm_obj)
+    except CliTableError:
+        return raw_output
+
+
+def clitable_to_dict(cli_table):
+    """Converts TextFSM cli_table object to list of dictionaries."""
+    objs = []
+    for row in cli_table:
+        temp_dict = {}
+        for index, element in enumerate(row):
+            temp_dict[cli_table.header[index].lower()] = element
+        objs.append(temp_dict)
+    return objs
+
+
+def netmiko_args(task, ip=None, host=None, username=None, password=None, device_type=None,
+                 netmiko_dict=None):
+    """Process arguments coming from different sources."""
+    parameters = {
+        "username": username or task.host["brigade_username"],
+        "password": password or task.host["brigade_password"],
+    }
+
+    if host is None and ip is None:
+        parameters["ip"] = task.host["brigade_ip"]
+    elif ip is not None:
+        parameters["ip"] = ip
+    elif host is not None:
+        parameters["host"] = host
+
+    if netmiko_dict is not None:
+        parameters.update(netmiko_dict)
+    else:
+        netmiko_dict = {}
+
+    if device_type is None:
+        if netmiko_dict.get("device_type") is None:
+            device_type = task.host["nos"]
+        else:
+            device_type = netmiko_dict.get("device_type")
+
+    # Convert to netmiko device_type format (if napalm format is used)
+    parameters['device_type'] = napalm_to_netmiko_map.get(device_type, device_type)
+    return parameters
+
+
+def netmiko_run(task, method, ip=None, host=None, username=None, password=None,
```

Comment:
I think we should simplify tasks and remove `ip=None, host=None, username=None, password=None, device_type=None` as those can come easily from the inventory. Or is there any reason why you think you'd want to override this information because it's easier than getting it from the inventory?

Reply:
My preference is to have both as it gives flexibility on how you run the functions...i.e. can be run potentially independently of inventory.

I don't care that much though.

Reply:
I am fine with having a mechanism to override parameters but then I'd suggest having a mechanism we can reuse everywhere instead of solving it on each task in an adhoc manner. For instance, we could accept a "connection_parameters" in all the tasks and have a helper function in the host itself that resolves the parameters.

Reply:
Yes, definitely open to figuring out some common pattern.